src/Remapping/remapping_utils.jl

```diff
@@ -0,0 +1,268 @@
+"""
+function vertical_reference_coordinates(space, zcoords)
+    vert_topology = Spaces.vertical_topology(space)
+    vert_mesh = vert_topology.mesh
+
+    is_cell_center = space isa Spaces.CenterExtrudedFiniteDifferenceSpace
+
+    ξ3s =
+        map(zcoords) do zcoord
+            velem = Meshes.containing_element(vert_mesh, zcoord)
+            ξ3, = Meshes.reference_coordinates(vert_mesh, velem, zcoord)
+            # For cell centered spaces, shift ξ3 so that we can use it for linear interpolation
+            is_cell_center && (ξ3 = ξ3 < 0 ? ξ3 + 1 : ξ3 - 1)
+            return ξ3
+        end |> collect
+
+    return ξ3s
+end
+
+
+"""
+    vertical_bounding_indices(space, zcoords)
+
+Return the vertical element indices needed to perform linear interpolation of `zcoords`.
+
+For centered-valued fields, if `zcoord` is in the top (bottom) half of a top (bottom)
+element in a column, no interpolation is performed and the value at the cell center is
+returned. Effectively, this means that the interpolation is first-order accurate across the
+column, but zeroth-order accurate close to the boundaries.
+"""
+function vertical_bounding_indices(space, zcoords) end
+
+function vertical_bounding_indices(
+    space::Spaces.FaceExtrudedFiniteDifferenceSpace,
+    zcoords,
+)
+    vert_topology = Spaces.vertical_topology(space)
+    vert_mesh = vert_topology.mesh
+    velems = Meshes.containing_element.(Ref(vert_mesh), zcoords)
+    return velems, velems .+ 1
+end
+
+function vertical_bounding_indices(
+    space::Spaces.CenterExtrudedFiniteDifferenceSpace,
+    zcoords,
+)
+    vert_topology = Spaces.vertical_topology(space)
+    vert_mesh = vert_topology.mesh
+    Nz = Spaces.nlevels(space)
+
+    is_periodic = Topologies.isperiodic(vert_topology)
+
+    vert_indices =
+        map(zcoords) do zcoord
+            velem = Meshes.containing_element(vert_mesh, zcoord)
+            ξ3, = Meshes.reference_coordinates(vert_mesh, velem, zcoord)
+            if ξ3 < 0
+                v_lo = is_periodic ? mod1(velem - 1, Nz) : max(velem - 1, 1)
+                v_hi = velem
+            else
+                v_lo = velem
+                v_hi = is_periodic ? mod1(velem + 1, Nz) : min(velem + 1, Nz)
+            end
+            return v_lo, v_hi
+        end |> collect
+
+    return vert_indices
+end
+
+
+"""
+
+TODO
+"""
+function vertical_interpolation_weights(space, zcoords)
+    ξs = vertical_reference_coordinates(space, zcoords)
+    return [((1 - ξ) / 2, (1 + ξ) / 2) for ξ in ξs]
+end
+
+
+"""
+
+TODO
+"""
+function set_interpolated_values!(remapper, field)
+    set_interpolated_values!(
+        remapper._local_interpolated_values,
+        field,
+        remapper._field_values,
+        remapper.local_indices,
+        remapper.interpolation_matrices,
+        remapper.vert_interpolation_weights,
+        remapper.vert_bounding_indices,
+    )
+end
+
+function set_interpolated_values!(
+    out,
```

Comment:
What is `out` here?

Reply:
The preallocated area of memory where we save the interpolated values. This is a private field in the remapper, used to assemble the result at the end.